File: src/main/java/com/example/domain/review/entity/Review.java
package com.example.domain.review.entity;

import com.example.domain.festival.entity.Festival;
import com.example.domain.member.entity.Member;
import com.example.global.entity.BaseEntity;
import jakarta.persistence.*;
import jakarta.validation.constraints.Max;
import jakarta.validation.constraints.Min;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Review extends BaseEntity {

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "member_id", nullable = false)
    private Member member;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "festival_id", nullable = false)
    private Festival festival;

    @Column(nullable = false, columnDefinition = "TEXT")
    private String content;

    private String image;

    @Min(value = 1, message = "평점은 최소 1점입니다.")
    @Max(value = 5, message = "평점은 최대 5점입니다.")
    @Column(nullable = false)
    private Integer rating;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private ReviewStatus status = ReviewStatus.ACTIVE;

    @Column(nullable = false)
    private Integer likeCount = 0;

    @Column(nullable = false)
    private Integer reportCount = 0;

    @Builder
    public Review(Member member, Festival festival, String content, String image, Integer rating) {
        this.member = member;
        this.festival = festival;
        this.content = content;
        this.image = image;
        this.rating = rating;
        this.status = ReviewStatus.ACTIVE;
        this.likeCount = 0;
        this.reportCount = 0;
    }


    public void updateReview(String content, String image, Integer rating) {
        this.content = content;
        this.image = image;
        this.rating = rating;
    }

    public void deleteReview() {
        this.status = ReviewStatus.DELETED;
    }

    public void reviewBlind(){
        this.status=ReviewStatus.BLIND;
    }

    public  void reportCountReset(){
        this.reportCount=0;
    }

    // 리뷰가 신고될 때마다 누적 신고 수를 1 증가시킵니다.
    public void increaseReportCount() {
        this.reportCount++;
    }
    //리뷰 좋아요 카운트 및 좋아요 취소
    public void increaseLikeCount() {
        this.likeCount++;
    }

    public void decreaseLikeCount() {
        if (this.likeCount > 0) {
            this.likeCount--;
        }
    }

}

File: src/main/kotlin/com/example/domain/review/entity/Review.kt
package com.example.domain.review.entity

import com.example.domain.festival.entity.Festival
import com.example.domain.member.entity.Member
import com.example.global.entity.BaseEntity
import jakarta.persistence.*
import jakarta.validation.constraints.Max
import jakarta.validation.constraints.Min

@Entity
@Table(name = "review")
class Review(
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "member_id", nullable = false)
    val member: Member,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "festival_id", nullable = false)
    val festival: Festival,

    @Column(nullable = false, columnDefinition = "TEXT")
    var content: String,

    var image: String?,

    @field:Min(value = 1, message = "평점은 최소 1점입니다.")
    @field:Max(value = 5, message = "평점은 최대 5점입니다.")
    @Column(nullable = false)
    var rating: Int

) : BaseEntity() {

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    var status: ReviewStatus = ReviewStatus.ACTIVE
        protected set

    @Column(nullable = false)
    var likeCount: Int = 0
        protected set

    @Column(nullable = false)
    var reportCount: Int = 0
        protected set

    fun updateReview(content: String, image: String?, rating: Int) {
        this.content = content
        this.image = image
        this.rating = rating
    }

    fun deleteReview() {
        this.status = ReviewStatus.DELETED
    }

    fun reviewBlind() {
        this.status = ReviewStatus.BLIND
    }

    fun reportCountReset() {
        this.reportCount = 0
    }

    // 리뷰가 신고될 때마다 누적 신고 수를 1 증가시킵니다.
    fun increaseReportCount() {
        this.reportCount++
    }

    //리뷰 좋아요 카운트 및 좋아요 취소
    fun increaseLikeCount() {
        this.likeCount++
    }

    fun decreaseLikeCount() {
        if (this.likeCount > 0) {
            this.likeCount--
        }
    }
}